File: clients/native-app/labs/lab05_dead_reckoning.c
/*
 * Lab 05 — Dead Reckoning.
 *
 * Instead of drawing the PAST (lerp), forward-simulate the latest snapshot to
 * the PRESENT with the same step function the server runs. The reckon horizon
 * is exactly the snapshot age.
 *
 *   smoothing  glide applied to each snapshot REBASE (the small correction
 *              when a patch lands mid-glide).
 *   snap       rebases beyond this distance POP instead of gliding: a teleport
 *              is a cut, and smoothing across it looks like flying.
 *
 * Port of src/client/labs/05-dead-reckoning/.
 */

#define L05_MAX_DOTS 40

static const char* const L05_PATTERNS[] = { "teleport", "patrol", "wander", "circle" };

static struct {
    colyseus_room_t* room;
    colyseus_room_clock_t* clock;
    colyseus_callbacks_t* callbacks;
    colyseus_predict_t* lerp;
    colyseus_predict_t* reckon;
    colyseus_input_handle_t* input;
    move_input_t* cmd;
    bots_state_t* state;
    const char* sid;
    bot_t* bot;

    pacer_t pacer;
    double smoothing, snap;
    int pattern;

    struct { double x, y, t; } dots[L05_MAX_DOTS];
    int dot_count;
    double last_raw_x, last_raw_y;
    double last_reckon_x;
    int warps;
    double warp_flash_t;
} l05;

/*
 * The reckon step, called by the SDK with a scratch copy of the bot instance.
 * `elapsed_ms` is the absolute server-time at the end of the substep, which is
 * what makes the circle's closed form and the teleport schedule evaluable at
 * any instant — the same property the server relies on.
 */
static void l05_reckon_step(colyseus_schema_t* state, double dt, double elapsed_ms, void* userdata) {
    (void)userdata;
    bot_t* b = (bot_t*)state;
    bot_sim_t s = {
        b->x, b->y, b->vx, b->vy, b->kind,
        b->minX, b->maxX, b->baseY, b->phaseMs, b->speed, b->lastTeleport,
    };
    step_bot(&s, dt, elapsed_ms);
    b->x = s.x; b->y = s.y; b->vx = s.vx; b->vy = s.vy;
    /* Scratch mutation only persists within one forward pass — exactly right:
     * multiple teleport periods chain, but nothing leaks into the snapshot. */
    b->lastTeleport = s.last_teleport;
}

static bool l05_attach_reckon(void) {
    static const char* const FIELDS[] = { "x", "y" };
    l05.reckon = colyseus_predict_create(l05.callbacks, l05.clock);
    return colyseus_predict_track_reckon(l05.reckon, (colyseus_schema_t*)l05.bot, &bot_vtable,
        FIELDS, 2, l05_reckon_step, l05.smoothing, 0, l05.snap, NULL) == 0;
}

static void l05_send_pattern(void) {
    colyseus_message_t* m = colyseus_message_map_create();
    colyseus_message_map_put_str(m, "kind", L05_PATTERNS[l05.pattern]);
    colyseus_room_send(l05.room, "pattern", m);
    colyseus_message_free(m);
}

static bool lab05_attach(app_t* app, colyseus_room_t* room) {
    bots_state_t* state = (bots_state_t*)colyseus_room_get_state(room);
    if (!state || !state->bots || !state->players) { return false; }
    const char* sid = colyseus_room_get_session_id(room);
    bot_t* bot = (bot_t*)colyseus_map_schema_get(state->bots, "bot1");
    if (!bot || !colyseus_map_schema_get(state->players, sid)) { return false; }

    memset(&l05, 0, sizeof(l05));
    l05.room = room;
    l05.state = state;
    l05.sid = sid;
    l05.bot = bot;
    l05.clock = colyseus_room_get_clock(room);
    l05.smoothing = 25;
    l05.snap = 8;
    l05.last_raw_x = NAN;
    l05.last_raw_y = NAN;
    l05.last_reckon_x = NAN;
    l05.warp_flash_t = -1e9;
    l05.callbacks = colyseus_room_callbacks(room);   /* room-owned, shared by the overlays */

    /* The delayed baseline to compare against. */
    l05.lerp = colyseus_predict_create(l05.callbacks, l05.clock);
    colyseus_predict_field_options_t lerp_opts = { 0 };
    lerp_opts.mode = COLYSEUS_PREDICT_LERP;
    lerp_opts.delay = REMOTE_INTERP_MS;
    colyseus_predict_track(l05.lerp, (colyseus_schema_t*)bot, "x", &lerp_opts);
    colyseus_predict_track(l05.lerp, (colyseus_schema_t*)bot, "y", &lerp_opts);

    if (!l05_attach_reckon()) { return false; }

    l05.input = colyseus_room_input(room, &move_input_vtable, NULL);
    if (!l05.input) { return false; }
    l05.cmd = (move_input_t*)colyseus_input_handle_data(l05.input);
    pacer_init(&l05.pacer, 1000.0 / TICK_HZ);

    l05_send_pattern();   /* the teleporter is the most legible default */
    (void)app;
    return true;
}

static void l05_draw_player(const char* sid, void* value, void* userdata) {
    app_t* app = (app_t*)userdata;
    player_t* p = (player_t*)value;
    draw_square(&app->view, p->x, p->y, PLAYER_HALF,
        hue_color(p->hue, strcmp(sid, l05.sid) == 0 ? 0.9 : 0.4));
}

static void l05_rebuild(void) {
    colyseus_predict_free(l05.reckon);
    l05_attach_reckon();
}

static void lab05_frame(app_t* app, double now, double dt) {
    (void)dt;
    /* --- controls ---------------------------------------------------- */
    if (app_key(KEY_B)) {
        l05.pattern = (l05.pattern + 1) % 4;
        l05_send_pattern();
        l05.warps = 0;
    }
    int smooth_step = app_key(KEY_EQUAL) ? 5 : app_key(KEY_MINUS) ? -5 : 0;
    if (smooth_step) {
        l05.smoothing += smooth_step;
        if (l05.smoothing < 0) { l05.smoothing = 0; }
        if (l05.smoothing > 50) { l05.smoothing = 50; }
        l05_rebuild();
    }
    int snap_step = app_key(KEY_PERIOD) ? 6 : app_key(KEY_COMMA) ? -6 : 0;
    if (snap_step) {
        l05.snap += snap_step;
        if (l05.snap < 1) { l05.snap = 1; }
        if (l05.snap > 60) { l05.snap = 60; }
        l05_rebuild();
    }

    int steps = pacer_steps(&l05.pacer, now);
    for (int i = 0; i < steps; i++) {
        l05.cmd->moveX = (int8_t)kb_move_x();
        l05.cmd->moveY = (int8_t)kb_move_y();
        colyseus_input_handle_send(l05.input);
    }
    colyseus_predict_tick(l05.lerp, now);
    colyseus_predict_tick(l05.reckon, now);

    double s_now = colyseus_room_clock_server_now(l05.clock);
    double age = fmax(0, s_now - colyseus_room_clock_last_server_time(l05.clock));
    const world_view_t* v = &app->view;

    /* Raw snapshot dots. */
    if (l05.bot->x != l05.last_raw_x || l05.bot->y != l05.last_raw_y) {
        l05.last_raw_x = l05.bot->x;
        l05.last_raw_y = l05.bot->y;
        if (l05.dot_count == L05_MAX_DOTS) {
            memmove(l05.dots, l05.dots + 1, sizeof(l05.dots[0]) * (L05_MAX_DOTS - 1));
            l05.dot_count--;
        }
        l05.dots[l05.dot_count].x = l05.bot->x;
        l05.dots[l05.dot_count].y = l05.bot->y;
        l05.dots[l05.dot_count].t = now;
        l05.dot_count++;
    }

    colyseus_map_schema_foreach(l05.state->players, l05_draw_player, app);

    for (int i = 0; i < l05.dot_count; i++) {
        double a = 1 - (now - l05.dots[i].t) / 1500.0;
        if (a <= 0) { continue; }
        draw_circle_world(v, l05.dots[i].x, l05.dots[i].y, 0.4, with_alpha(COL_TEXT, a * 0.5));
    }

    double lx = colyseus_predict_value(l05.lerp, (colyseus_schema_t*)l05.bot, "x");
    double ly = colyseus_predict_value(l05.lerp, (colyseus_schema_t*)l05.bot, "y");
    double rx = colyseus_predict_value(l05.reckon, (colyseus_schema_t*)l05.bot, "x");
    double ry = colyseus_predict_value(l05.reckon, (colyseus_schema_t*)l05.bot, "y");

    if (!isnan(l05.last_reckon_x) && fabs(rx - l05.last_reckon_x) > 15) {
        l05.warps++;
        l05.warp_flash_t = now;
    }
    l05.last_reckon_x = rx;

    /* The reckon horizon: newest snapshot -> forward-simulated present. */
    draw_dashed_world(v, l05.bot->x, l05.bot->y, rx, ry, with_alpha(COL_WARN, 0.8), 1.2f);
    draw_circle_world(v, l05.bot->x, l05.bot->y, 0.7, with_alpha(COL_TEXT, 0.9));

    draw_circle_outline_world(v, lx, ly, BOT_RADIUS, COL_BLUE);
    draw_label_world(v, lx, ly, "lerp (past)", COL_BLUE, 10, vs(v, BOT_RADIUS) + 4);
    draw_circle_world(v, rx, ry, BOT_RADIUS, with_alpha(COL_WARN, 0.25));
    draw_circle_outline_world(v, rx, ry, BOT_RADIUS, COL_WARN);
    draw_label_world(v, rx, ry, "reckon (present)", COL_WARN, 10, -vs(v, BOT_RADIUS) - 14);

    if (now - l05.warp_flash_t < 500) {
        draw_label_world(v, rx, ry, "WARP", COL_BAD, 13, -vs(v, BOT_RADIUS) - 30);
    }

    /* --- HUD --------------------------------------------------------- */
    hud_t* h = &app->hud;
    hud_section(h, "TELEMETRY");
    hud_row(h, "reckon horizon (snapshot age)", COL_TEXT, "%.0f ms", age);
    hud_row(h, "reckon vs lerp gap", COL_TEXT, "%.1f u",
        sqrt((rx - lx) * (rx - lx) + (ry - ly) * (ry - ly)));
    hud_row(h, "warps seen", COL_TEXT, "%d", l05.warps);

    hud_section(h, "CONTROLS");
    hud_key(h, "WASD", "drive your own square");
    hud_key(h, "B", TextFormat("bot pattern: %s", L05_PATTERNS[l05.pattern]));
    hud_key(h, "- / =", TextFormat("rebase smoothing  %.0f /s", l05.smoothing));
    hud_key(h, ", / .", TextFormat("snap threshold  %.0f u", l05.snap));
    hud_note(h, "patrol = fully predictable - wander = server-secret turns, so reckon "
        "extrapolates straight through every one and gets corrected - teleport = a "
        "scheduled discontinuity. Raise the snap threshold above the warp distance and "
        "watch the teleport smear across the arena.");
}

static void lab05_detach(app_t* app) {
    (void)app;
    if (l05.reckon) { colyseus_predict_free(l05.reckon); }
    if (l05.lerp) { colyseus_predict_free(l05.lerp); }
    memset(&l05, 0, sizeof(l05));
}

static void lab05_on_reconnect(app_t* app) {
    (void)app;
    pacer_reset(&l05.pacer);
    l05.bot = (bot_t*)colyseus_map_schema_get(l05.state->bots, "bot1");
    l05.dot_count = 0;
}

static const lab_def_t LAB_05 = {
    .id = "05-dead-reckoning",
    .num = 5,
    .title = "Dead Reckoning",
    .blurb = "Forward-simulate remotes to the present with the shared step.",
    .room_name = "lab-bots",
    .state_vtable = &bots_state_vtable,
    .attach = lab05_attach,
    .frame = lab05_frame,
    .detach = lab05_detach,
    .on_reconnect = lab05_on_reconnect,
};

File: clients/flutter-app/lib/labs/lab05_dead_reckoning.dart
import 'dart:math' as math;

import 'package:colyseus_flutter/colyseus_flutter.dart';

import '../controls.dart';
import '../kb.dart';
import '../lab.dart';
import '../net/net_delay.dart';
import '../net/schema_bridge.dart';
import '../pacer.dart';
import '../palette.dart';
import '../sim/sim.dart';

/// Drawing the present instead of the past.
///
/// Interpolation renders between two samples that already arrived, so it is
/// always a delay behind. Dead reckoning takes the newest snapshot and runs it
/// forward with the same step function the server runs — [stepBot] out of
/// `lib/sim`, the very code the server's TypeScript is a twin of. The distance
/// between the two ghosts is the delay that buys back.
///
/// The dashed line from the white dot is the reckon horizon: it grows between
/// patches and collapses when one lands, so its length is the snapshot age.
///
/// What to look for per pattern:
/// - patrol and circle are entirely in the shared step, so the reckoned bot
///   sits on the truth and corrections stay at wire precision.
/// - teleport rides a synced schedule (`lastTeleport`), so even the warp is
///   predicted; the snap threshold decides whether the residual pops or glides.
/// - wander re-rolls its heading from a seed only the server has. Reckoning
///   runs straight through every turn and gets corrected. That error is the
///   point, not a defect.
class Lab05DeadReckoning extends Lab {
  @override
  String get id => '05';

  @override
  String get title => 'Dead Reckoning';

  @override
  String get blurb => 'Forward-simulate the last snapshot with the shared step.';

  static const _patterns = ['patrol', 'circle', 'wander', 'teleport'];
  static const _maxDots = 40;

  /// An x jump this large in one frame is a warp, not motion.
  static const _warpDistance = 15.0;

  Predict? _lerp;
  Predict? _reckon;
  InputHandle? _input;
  final _pacer = FixedStepPacer(1000 / tickHz);
  final _binding = StepBinding();

  final List<({double x, double y, double t})> _dots = [];

  String _pattern = 'patrol';
  double _smoothing = 25;
  double _snap = 8;

  /// The decoded `kind`, handed to the step: the reckon scratch carries
  /// numbers only, so a string field does not survive into it.
  String _kind = '';

  double _lastRawX = double.nan;
  double _lastRawY = double.nan;
  double _lastReckonX = double.nan;
  double _warpFlashAt = double.negativeInfinity;
  int _warps = 0;

  @override
  Future<bool> mount(LabContext ctx) async {
    final joined = await ctx.client.joinOrCreate('lab-bots');
    room = joined;
    NetDelay.register(joined);

    _input = joined.input();
    if (_input == null) return false;
    if (await _waitForBot() == null) return false;

    // The delayed baseline to measure the reckoned one against.
    _lerp = Predict.of(joined)
      ..attachAll('bots', config: {
        'x': const FieldOptions(mode: PredictMode.lerp, delay: remoteInterpMs),
        'y': const FieldOptions(mode: PredictMode.lerp, delay: remoteInterpMs),
      });

    _attachReckon();
    _send();
    return true;
  }

  /// The bot, re-read every frame — the decoder can replace instances.
  SchemaInstance? get bot =>
      room?.state?.getMap('bots')?['bot1'] as SchemaInstance?;

  Future<SchemaInstance?> _waitForBot() async {
    final deadline = DateTime.now().add(const Duration(seconds: 5));
    while (DateTime.now().isBefore(deadline)) {
      final found = bot;
      if (found != null) return found;
      await Future<void>.delayed(const Duration(milliseconds: 16));
    }
    return null;
  }

  void _attachReckon() {
    final joined = room;
    if (joined == null) return;

    _reckon?.dispose();
    _reckon = Predict.of(joined)
      ..attachAllReckon('bots', const ['x', 'y'], _step,
          smoothing: _smoothing, snap: _snap);
  }

  /// The forward pass, called by the SDK with a scratch copy of the snapshot.
  ///
  /// [elapsedMs] is absolute server time at the end of the sub-step, which is
  /// what makes the circle's closed form and the teleport schedule evaluable
  /// at any instant rather than only on tick boundaries.
  void _step(SchemaView state, double dt, double elapsedMs) {
    stepBot(_binding.body(state), dt, elapsedMs, kindOverride: _kind);
  }

  void _send() => room?.send('pattern', {'kind': _pattern});

  @override
  void frame(LabContext ctx, double now, double dtMs) {
    final joined = room;
    final input = _input;
    final lerp = _lerp;
    final reckon = _reckon;
    if (joined == null || input == null || lerp == null || reckon == null) {
      return;
    }

    // No reconciler here, so the sends are paced on a timer; they are also
    // what keeps the clock estimate fed.
    final steps = _pacer.steps(now);
    for (var i = 0; i < steps; i++) {
      input.data['moveX'] = Kb.moveX();
      input.data['moveY'] = Kb.moveY();
      input.send();
    }

    lerp.tick(now);
    reckon.tick(now);

    final target = bot;
    if (target == null) return;
    _kind = (target['kind'] as String?) ?? '';

    final rawX = (target['x'] as num).toDouble();
    final rawY = (target['y'] as num).toDouble();
    if (rawX != _lastRawX || rawY != _lastRawY) {
      _lastRawX = rawX;
      _lastRawY = rawY;
      if (_dots.length == _maxDots) _dots.removeAt(0);
      _dots.add((x: rawX, y: rawY, t: now));
    }

    final reckonX = reckon.value(target, 'x');
    if (!_lastReckonX.isNaN && (reckonX - _lastReckonX).abs() > _warpDistance) {
      _warps++;
      _warpFlashAt = now;
    }
    _lastReckonX = reckonX;
  }

  @override
  void render(LabContext ctx) {
    final joined = room;
    final lerp = _lerp;
    final reckon = _reckon;
    if (joined == null || lerp == null || reckon == null) return;

    final draw = ctx.draw;
    final now = joined.clock.now;

    final players = joined.state?.getMap('players');
    if (players != null) {
      for (final entry in players.entries) {
        final instance = entry.value;
        if (instance is! SchemaInstance) continue;
        draw.square(
          (instance['x'] as num).toDouble(),
          (instance['y'] as num).toDouble(),
          playerHalf,
          hueColor((instance['hue'] as num?)?.toInt() ?? 0,
              alpha: entry.key == joined.sessionId ? 0.9 : 0.4),
        );
      }
    }

    // The snapshots the forward pass starts from, fading over 1.5 s.
    for (final dot in _dots) {
      final alpha = 1 - (now - dot.t) / 1500;
      if (alpha <= 0) continue;
      draw.circle(dot.x, dot.y, 0.4, Palette.text.fade(alpha * 0.5));
    }

    final target = bot;
    if (target == null) return;

    final rawX = (target['x'] as num).toDouble();
    final rawY = (target['y'] as num).toDouble();
    final lerpX = lerp.value(target, 'x');
    final lerpY = lerp.value(target, 'y');
    final reckonX = reckon.value(target, 'x');
    final reckonY = reckon.value(target, 'y');

    // The horizon: newest snapshot to the forward-simulated present.
    draw.dashedLine(rawX, rawY, reckonX, reckonY, Palette.warn.fade(0.8),
        width: 1.2);
    draw.circle(rawX, rawY, 0.7, Palette.text.fade(0.9));

    draw.circleOutline(lerpX, lerpY, botRadius, Palette.blue, width: 2);
    draw.label(lerpX, lerpY, 'lerp (past)', Palette.blue,
        size: 10, dy: ctx.view.s(botRadius) + 4);

    draw.circle(reckonX, reckonY, botRadius, Palette.warn.fade(0.25));
    draw.circleOutline(reckonX, reckonY, botRadius, Palette.warn, width: 2);
    draw.label(reckonX, reckonY, 'reckon (present)', Palette.warn,
        size: 10, dy: -ctx.view.s(botRadius) - 14);

    if (now - _warpFlashAt < 500) {
      draw.label(reckonX, reckonY, 'WARP', Palette.bad,
          size: 13, dy: -ctx.view.s(botRadius) - 30);
    }

    _renderHud(ctx, lerpX, lerpY, reckonX, reckonY);
  }

  void _renderHud(
    LabContext ctx,
    double lerpX,
    double lerpY,
    double reckonX,
    double reckonY,
  ) {
    final hud = ctx.hud;
    final clock = room!.clock;
    final age = math.max(0.0, clock.serverNow - clock.lastServerTime);
    final gap = math.sqrt((reckonX - lerpX) * (reckonX - lerpX) +
        (reckonY - lerpY) * (reckonY - lerpY));

    hud.section('RECKONING');
    hud.row('horizon (snapshot age)', '${age.round()} ms');
    hud.row('reckon vs lerp gap', '${gap.toStringAsFixed(1)} u');
    hud.row('warps seen', '$_warps');
    hud.row('pattern', _pattern);
    hud.row('injected', NetDelay.current.label);

    hud.section('CLOCK');
    hud.row('rtt', '${clock.smoothedRtt.round()} ms');
    hud.row('patch interval', '${clock.patchInterval.round()} ms');

    hud.note(switch (_pattern) {
      'wander' => 'wander re-rolls its heading from a seed only the server '
          'has. The forward pass runs straight through every turn and is '
          'corrected on the next patch, which is why the gap keeps reopening.',
      'teleport' => 'The warp schedule is synced state, so the forward pass '
          'predicts the discontinuity too. Raise the snap threshold past the '
          'warp distance and the correction glides across the arena instead '
          'of cutting.',
      _ => 'This pattern lives entirely in the shared step, so the reckoned '
          'ring sits on the truth and the horizon line is the only thing that '
          'moves between patches.',
    });
  }

  @override
  List<ControlSpec> controls(LabContext ctx) {
    return [
      RadioSpec<String>(
        'bot pattern (room-wide)',
        [for (final kind in _patterns) (value: kind, label: kind)],
        _pattern,
        (kind) {
          _pattern = kind;
          _send();
          _warps = 0;
          _dots.clear();
        },
      ),
      SliderSpec('rebase smoothing', 0, 50, _smoothing, (v) {
        if (v == _smoothing) return;
        _smoothing = v;
        _attachReckon();
      }, divisions: 10, format: (v) => '${v.round()} /s'),
      SliderSpec('snap threshold', 2, 62, _snap, (v) {
        if (v == _snap) return;
        _snap = v;
        _attachReckon();
      }, divisions: 10, format: (v) => '${v.round()} u'),
      const NoteSpec('Smoothing is how fast a snapshot rebase glides out. '
          'Anything past the snap threshold cuts instead, because smoothing '
          'across a teleport draws the bot crossing ground it never covered.'),
    ];
  }

  @override
  void onReconnect() {
    _pacer.reset();
    _dots.clear();
    _lastRawX = double.nan;
    _lastRawY = double.nan;
    _lastReckonX = double.nan;
  }

  @override
  void unmount() {
    _reckon?.dispose();
    _reckon = null;
    _lerp?.dispose();
    _lerp = null;
    _dots.clear();
    NetDelay.unregister(room!);
  }
}
